hcl2template/types.packer_config.go

```diff
@@ -238,23 +240,41 @@ func (c *PackerConfig) evaluateLocalVariables(locals []*LocalBlock) hcl.Diagnost
 		c.LocalVariables = Variables{}
 	}
 
-	for foundSomething := true; foundSomething; {
-		foundSomething = false
-		for i := 0; i < len(locals); {
-			local := locals[i]
-			moreDiags := c.evaluateLocalVariable(local)
-			if moreDiags.HasErrors() {
-				i++
+	for _, local := range c.LocalBlocks {
+		// Note: when looking at the expressions, we only need to care about
+		// attributes, as HCL2 expressions are not allowed in a block's labels.
+		vars := FilterTraversalsByType(local.Expr.Variables(), "local")
+
+		var localDeps []*LocalBlock
+		for _, v := range vars {
+			// Some local variables may be locally aliased as
```

Comment:
Doing a quick test I see the ability to define the same local twice as a potential issue. 

```hcl
packer {
    required_plugins {
        gcp = {
        source = "github.com/hashicorp/googlecompute"
        version = "=1.0.13"
        }
        gcp2 = {
        source = "github.com/hashicorp/googlecompute"
        version = "=1.0.12"
        }
    }
}

local "test" {
    expression = "two"
    sensitive = true
}

locals {
 test = local.test
}


variable "test" {
    type = string
    default = "home"
}
source "null" "example" {}

build {
sources = ["source.null.example"]
}
```
The sensitive flag is dropped in this case
```shell
~>  packer console source.pkr.hcl
> variables
> > input-variables:

var.test: "home"

> local-variables:

local.test: "two"
```

Reply:
Ah no that's not the same thing, we do have a check that rejects duplicate definitions of a variable, so that template would fail (I think, need to double check, but I would expect that).

That check here ensures we don't have something that is called `local` in a scope, there's one test template we have where that happens:

```hcl

variable "vms_to_build" {
}

locals {
  vms_to_build = var.vms_to_build
  dynamic_map = { 
    for vm in local.vms_to_build : 
      vm => lookup(local, vm, "VM NAME NOT FOUND")
  }
}
```

Here in the context of the `dynamic_map`, we have a symbol called `local`, which doesn't match an existing variable for example, and since there's not multiple parts to the name, we just ignore it.

Maybe my comment isn't that clear though, any idea how we can rephrase that to capture the intent?

Reply:
If we can define the same variable twice though, that's a red flag; we should definitely error in that case.
I thought we had a check for this, but maybe I broke it with that change, will follow-up.

Reply:
Note: as discussed out-of-band, the problem here was that the duplicate detection was done during `Initialise`, so while that was reported correctly for commands like `build` or `validate`, `console` did not as the returned diagnostics from `Initialise` were ignored.
We moved that logic to during parsing since the names are static, so that the check is always performed and honoured, regardless of the command.